### PR TITLE
Datastructures: Stacks & Linked Lists

### PR DESCRIPTION
**Pull request by RemieRichards**
## General changes:
- Adds the /__DATASTRUCTURES folder
- Moves PriorityQueue data structure out of the AStar file, and into it's own file in /__DATASTRUCTURES
- Adds Stacks, a data structure, to /__DATASTRUCTURES
- Adds Linked Lists, a data structure, to /__DATASTRUCTURES
- Deletes `stack.dm` and the merges the implementation of NTSL's stacks into this version of stacks (for organisational sake and because mine had more procs and features)
## Stack:

A stack is a First-in-Last-Out data structure, Elements are accessed with the `pop()` and `push()` methods primarily, `pop()` returns the last element in the stack, and removes it from the stack (there is also `peek()` which returns the last element without removing it), `push()` adds the element on to the end of the stack, making it the element returned by the next `pop()`
## Linked List:

Linked List: a double-linked linked list, which stores `linked_nodes`s which contain a reference to the node before them (or null), the node after them (or null), the `linked_list` they belong to and the value they hold. Thanks to @Giacom and @MrStonedOne for pointing out the oddity of my original implementation.

I was bored so I implemented Stacks and Linked Lists (double linked lists) in Byond, woo.
## Performance:

<B>1 Million calls</B> to <B>Stack Pop()</B>, was around <B>1CPU</B>
<B>1 Million calls</B> to <B>Stack Push()</B>, was around <B>4CPU</B>
<B>1 Million calls</B> to <B>Linked List Add()</B>, was around <B>3CPU</B>
<B>1 Million calls</B> to <B>Linked List Remove()</B>, was around <B>4CPU</B>

### CODE REVIEW

**Comment by MrPerson** [comment]
What possible use do linked lists have that couldn't be replicated with regular byond lists?

**Comment by RemieRichards** [comment]
@MrPerson never said they'd be useful, I'm just offering them up since I spent 2 hours debugging them.

**Comment by MrStonedOne** [comment]
linked lists gives you an ability to send a list to a proc but starting at a certain point, and it can just keep using *next

File: code/__DATASTRUCTURES/stacks.dm [hunk 1]
```diff
@@ -0,0 +1,55 @@
+
+//Both count as failures, and they don't equate to each other
+//this lets us do if(pop()) without having to specifically check for underflow
+//same for if(push()) and overflow
+#define STACK_OVERFLOW	""
+#define STACK_UNDERFLOW	null
+
+
+/datum/stack
```
Comment (Iamgoofball) [inline code/__DATASTRUCTURES/stacks.dm]: somewhere sometime someone is going to make a datum related to material stacks named this, and get very confused why shit breaks

File: code/__DATASTRUCTURES/priority_queue.dm [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	var/i
+	L.Add(A)
+	i = L.len -1
+	while(i > 0 &&  call(cmp)(L[i],A) >= 0) //place the element at it's right position using the compare proc
+		L.Swap(i,i+1) 						//last inserted element being first in case of ties (optimization)
+		i--
+
+//removes and returns the first element in the queue
+/PriorityQueue/proc/Dequeue()
+	if(!L.len)
```
Comment (duncathan) [inline code/__DATASTRUCTURES/priority_queue.dm]: there is a proc that does this 14 lines above this
is this not what it would be used for???

**Comment by Miauw62** [comment]
I have a suspicion that these are going to be horribly, horribly slow (because of BYOND)

**Comment by Chnkr** [comment]
@Miauw62 i had the same suspicion, especially the stack, and the linked lists, byond is horrible at lists, and especially adding and removing things from a list.

File: code/__DATASTRUCTURES/priority_queue.dm (hunk 2 above)
Comment (vista-) [inline code/__DATASTRUCTURES/priority_queue.dm]: Making an extra function call is a performance trade-off with little to no gain in this case.

**Comment by RemieRichards** [comment]
@Miauw62 
@Chnkr 

1 million calls to Stack ```Pop()```, was around 3CPU
1 million calls to Stack ```Push()```, was around 7CPU
1 Million calls to Linked List ```add()```, was around 20CPU
1 Million calls to Linked List ```remove()```, was around 10CPU

The only way you're going to lag things is if you make stupid code, which is no different from how things are now.

@Chnkr as to your specific point on removing things, I was able to get ALL stack related removals to use ```Cut(i,i+1)``` which is the fastest method (fastest random access method) for removing an element from a list.  Sadly LL DO use the normal list ```Remove()``` instead of ```Cut()```, simply because I don't have the index immediately available, and I suspect a ```Find()``` call to get that Index would just defeat the whole purpose.

tl;dr everything is fine.

File: code/__DATASTRUCTURES/stacks.dm (hunk 1 above)
Comment (RemieRichards) [inline code/__DATASTRUCTURES/stacks.dm]: Tbh, Datastructures > Material stacks (that logically should just get called ```material_stack``` anyways :P)

**Comment by MrStonedOne** [comment]
Its kinda odd to do a linked list that is also stored as an array

Ideally the contents isn't needed other than to make it work with a for, but you lose performance gains in linked lists by doing that. and at that point its just an array with extra overhead.

File: code/__DATASTRUCTURES/linked_lists.dm [hunk 3]
```diff
@@ -0,0 +1,173 @@
+
+//Ok so it's technically a double linked list, bite me.
+
+/datum/linked_list
+	var/list/contents = list()
```
Comment (Giacom) [inline code/__DATASTRUCTURES/linked_lists.dm]: This isn't a real linked list, it is more like a [ring buffer](http://gameprogrammingpatterns.com/event-queue.html#a-ring-buffer) with the front and back pointers.

A real linked list would not have an underlying array (list), it would have nodes which point to nodes; while keeping track of the front and (maybe) back node.

File: code/__DATASTRUCTURES/priority_queue.dm [hunk 4]
```diff
@@ -0,0 +1,65 @@
+
+//////////////////////
+//PriorityQueue object
+//////////////////////
+
+//an ordered list, using the cmp proc to weight the list elements
+/PriorityQueue
+	var/list/L //the actual queue
```
Comment (Giacom) [inline code/__DATASTRUCTURES/priority_queue.dm]: Ideally a queue would use a linked list for the benefits of being able to remove/add the front element with O(1) efficiency.

Not sure how that would work in BYOND, you will need to do benchmarks. I am also assuming BYOND lists are dynamic arrays, they could be linked lists for all I know.

File: code/__DATASTRUCTURES/linked_lists.dm (hunk 3 above)
Comment (RemieRichards) [inline code/__DATASTRUCTURES/linked_lists.dm]: I can do that, gives me an excuse to redo the Front/Back naming scheme to.

File: code/__DATASTRUCTURES/priority_queue.dm [hunk 5]
```diff
@@ -0,0 +1,65 @@
+
+//////////////////////
+//PriorityQueue object
+//////////////////////
+
+//an ordered list, using the cmp proc to weight the list elements
+/PriorityQueue
```
Comment (Giacom) [inline code/__DATASTRUCTURES/priority_queue.dm]: Ideally a queue would use a linked list for the benefits of being able to remove/add the front element with O(1) efficiency.

Not sure how that would work in BYOND, you will need to do benchmarks. I am also assuming BYOND lists are dynamic arrays, they could be linked lists for all I know.

Comment (RemieRichards) [inline code/__DATASTRUCTURES/priority_queue.dm]: I'll look into that for a future PR, I'm not fully familiar with Priority Queues personally, just moving their contents to the (now) best place, the datastructure folder.

File: code/__DATASTRUCTURES/stacks.dm [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	if(elements)
+		stack = elements.Copy()
+	if(max)
+		max_elements = max
+
+/datum/stack/proc/pop()
+	if(is_empty())
+		return STACK_UNDERFLOW
+	. = stack[stack.len]
+	stack.Cut(stack.len,0)
```
Comment (Giacom) [inline code/__DATASTRUCTURES/stacks.dm]: Not sure if it is better but you can do this instead:

```C++
stack.len -= 1
```

File: code/__DATASTRUCTURES/stacks.dm [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+		return STACK_OVERFLOW
+	stack += element
+
+/datum/stack/proc/peek()
+	if(is_empty())
+		return STACK_UNDERFLOW
+	. = stack[stack.len]
+
+/datum/stack/proc/is_empty()
+	. = stack.len ? 0 : 1
```
Comment (Giacom) [inline code/__DATASTRUCTURES/stacks.dm]: Probably more clear to do

```C++
return stack.len > 0
```

File: code/__DATASTRUCTURES/stacks.dm (hunk 6 above)
Comment (RemieRichards) [inline code/__DATASTRUCTURES/stacks.dm]: P.sure there have been troubles in the past of ```len--`` or equivalent syntax failing to actually destroy the element.
```Cut()``` itself is faster than ```Remove()```, so it's decently fast already.

File: code/__DATASTRUCTURES/stacks.dm (hunk 7 above)
Comment (RemieRichards) [inline code/__DATASTRUCTURES/stacks.dm]: I went for speed here, no idea if ```>``` evaluates faster than ```?``` but ```.``` is certainly slightly faster than ```return```

File: code/__DATASTRUCTURES/linked_lists.dm (hunk 3 above)
Comment (Miauw62) [inline code/__DATASTRUCTURES/linked_lists.dm]: to the best of our knowledge, byond lists are red-black trees combined with something else for associativity.

Comment (RemieRichards) [inline code/__DATASTRUCTURES/linked_lists.dm]: @Miauw62 R/Bs+Dynamic(expanding) Arrays, R/Bs are only used for associations though, p.sure.

**Comment by RemieRichards** [comment]
Reimplemented Linked Lists per @Giacom and @MrStonedOne 's comments
CPU for Linked Lists has dropped from 20/10 to 4/4 on Add/Remove which is an added bonus.

File: code/__DATASTRUCTURES/stacks.dm (hunk 1 above)
Comment (PJB3005) [inline code/__DATASTRUCTURES/stacks.dm]: There already ARE stack datums, actually.

[honkhonkhonkhonk](https://github.com/tgstation/-tg-station/blob/master/code/modules/scripting/stack.dm)

File: code/__DATASTRUCTURES/linked_lists.dm [hunk 8]
```diff
@@ -0,0 +1,185 @@
+
+//Ok so it's technically a double linked list, bite me.
+
+/datum/linked_list
+	var/datum/linked_element/head
```
Comment (Giacom) [inline code/__DATASTRUCTURES/linked_lists.dm]: I think the generally agreed upon term is "node".

File: code/__DATASTRUCTURES/linked_lists.dm [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+/datum/linked_element
+	var/value = null
+	var/datum/linked_list/linked_list = null
+	var/datum/linked_element/next_element_pointer = null
+	var/datum/linked_element/previous_element_pointer = null
+
+
+/datum/linked_list/proc/is_empty()
+	. = element_amt ? 0 : 1
```
Comment (Giacom) [inline code/__DATASTRUCTURES/linked_lists.dm]:
```C++
. = element_amt <= 0
```

Makes more sense.

File: code/__DATASTRUCTURES/stacks.dm [hunk 10]
```diff
@@ -0,0 +1,55 @@
+
+//Both count as failures, and they don't equate to each other
+//this lets us do if(pop()) without having to specifically check for underflow
+//same for if(push()) and overflow
+#define STACK_OVERFLOW	""
```
Comment (Giacom) [inline code/__DATASTRUCTURES/stacks.dm]: I think it would be clearer if you made the DEFINEs into negative numbers. They would still count as false and it wouldn't be as confusing.

**Comment by Giacom** [comment]
If you or anyone else wants to implement more algorithms, or look through them to see which could help with an upcoming project, take a look at this.

https://github.com/aalhour/C-Sharp-Algorithms

It is in C# and there are a lot of generics and interfaces but you can strip them out.

**Comment by MrStonedOne** [comment]
Just a note, since its buried in a diff commit, this name conflicts with the stack datum used by ntsl.

Odds are, ntsl can be made to use this stack datum

**Comment by RemieRichards** [comment]
@MrStonedOne If I rename my Push() and Pop() to be capitalised, and implement NTSL stack's Copy() they'd be 100% compatible (Mine has some extra functions so mine would absorb the NTSL one, not to mention NTSL is generally considered kill these days)

@Giacom thanks, I'll keep a note of it.

File: code/__DATASTRUCTURES/stacks.dm (hunk 10 above)
Comment (RemieRichards) [inline code/__DATASTRUCTURES/stacks.dm]: That's true yes, at the time I forgot negative values equated to false

**Comment by RemieRichards** [comment]
News just in, Relative pathing still a shit.
I had to correct all the ```/stack``` to ```/datum/stack``` (despite them being the same) due to Relative pathing not recognising that /stack is shorthand for /datum/stack,  *sigh*

* ```/datum/stack``` had absorbed and assumed the job of ```/stack```
* Rewrote the naming scheme of ```linked_list``` to use ```linked_node```s instead of ```linked_element```s
* Changed the case for all the data structures to match, primarily because it was required for NTSL and also because I couldn't stand them not being uniform

**Comment by Giacom** [comment]
You should look into whether BYOND utilizes caching with its bytecode. Linked lists are not really favoured due to the fact the CPU can fetch batches of data from RAM; this can cause issue with linked lists since their location in RAM is not guaranteed to be next to each other, while arrays and other data structures which have an underlying array will guarantee it, at the cost of having to allocate another array when the existing one gets too small.

**Comment by MrStonedOne** [comment]
>You should look into whether BYOND utilizes caching with its bytecode. Linked lists are not really favoured due to the fact the CPU can fetch batches of data from RAM;

Tobba once said to me that not only did it not, the way it accessed memory made for extremely poor caching performance for what little caching could still be done by the cpu.

**Comment by Chnkr** [comment]
> You should look into whether BYOND utilizes caching with its bytecode

It does not, and as @MrStonedOne explained earlier the way it handles lists() internally makes for extremely poor performance during runtime, i suggest you implement this somewhere in-game, and profile it that way.

**Comment by Giacom** [comment]
Well that sucks.

**Comment by Razharas** [comment]
Will you use it for something in the end rr?
Also please dont let the usual fate get to this pr

**Comment by RemieRichards** [comment]
@Razharas WHAT USUAL FATE? Besides things in byond I can't control, it's fine to merge, has been for days.
